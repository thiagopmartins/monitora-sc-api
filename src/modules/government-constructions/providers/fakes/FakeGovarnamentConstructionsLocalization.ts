import GovernmentConstructionsLocalizations from '@modules/government-constructions/infra/typeorm/schemas/GovernmentConstructionsLocalizations';
import AppError from '@shared/errors/AppError';
import IGovernmentConstructionLocalizationProvider from '../IGovernmentConstructionLocalizationProvider';

class FakeGovarnamentConstructionsLocalization
  implements IGovernmentConstructionLocalizationProvider {
  private constructions: GovernmentConstructionsLocalizations[] = [];

  public async getAll(): Promise<
    GovernmentConstructionsLocalizations[] | undefined
  > {
    try {
      const construction = {
        id: 'SC110 _24',
        description:
          'PETROLÃ‚NDIA (ENTR. SC-284) - ENTR. BR-282 (P/ BOCAINA DO SUL)',
        nuLatitude: -27.535722175999,
        nuLongitude: -49.698887383295,
        urlPin:
          'http://www.sicop.sc.gov.br:80/mapa-smo-backend/imagens/areas/345pin.png',
        heatMapValue: 254912.79,
        coordenates: [],
        projectsId: [2239, 2259, 3992, 7192],
        contracts: [
          {
            id: 1026464,
            constructionSitesId: [1],
          },
          {
            id: 1027889,
            constructionSitesId: [1],
          },
          {
            id: 1029711,
            constructionSitesId: [1],
          },
        ],
        heatMapQuantity: 2.12,
      };

      const governamentConstructionsLocalizations = new GovernmentConstructionsLocalizations();

      Object.assign(
        governamentConstructionsLocalizations,
        { construction_id: construction.id },
        construction,
      );

      delete governamentConstructionsLocalizations.id;

      this.constructions.push(governamentConstructionsLocalizations);

      return this.constructions;
    } catch (error) {
      throw new AppError(error);
    }
  }
}

export default FakeGovarnamentConstructionsLocalization;
