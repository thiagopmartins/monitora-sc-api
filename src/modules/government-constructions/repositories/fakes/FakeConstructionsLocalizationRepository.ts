import ICreateConstructionsLocalizationDTO from '@modules/constructions/dtos/ICreateConstructionsLocalizationDTO';
import IConstructionsLocalization from '@modules/constructions/infra/mongoose/entities/IConstructionsLocalization';
import { ConstructionLocalizationType } from '@modules/constructions/infra/mongoose/schemas/ConstructionLocalizationSchema';
import { uuid } from 'uuidv4';
import IConstructionsLocalizationRepository from '../IConstructionsLocalizationRepository';

class FakeConstructionsLocalizationRepository
  implements IConstructionsLocalizationRepository {
  private constructions: IConstructionsLocalization[] = [];

  public async create(
    construction: ICreateConstructionsLocalizationDTO,
  ): Promise<ConstructionLocalizationType> {
    const teste: ConstructionLocalizationType = {};
    Object.assign(teste, { id: uuid() }, construction);
    this.constructions.push(construction);

    return teste;
  }

  public async findByConstructionLocalizarionId(
    id: string,
  ): Promise<IConstructionsLocalization | undefined> {
    const findConstruction = this.constructions.find(
      construction => construction.id === id,
    );
    return findConstruction;
  }
}

export default FakeConstructionsLocalizationRepository;
