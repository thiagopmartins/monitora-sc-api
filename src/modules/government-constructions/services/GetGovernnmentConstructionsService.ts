import { inject, injectable } from 'tsyringe';
import ICreateConstructionsLocalizationDTO from '../dtos/ICreateConstructionsLocalizationDTO';
import IGovernmentConstructionLocalizationProvider from '../providers/IGovernmentConstructionLocalizationProvider';
import IGovernmentConstructionsLocalizationsRepository from '../repositories/IGovernmentConstructionsLocalizationsRepository';

@injectable()
class GetGovernnmentConstructionsService {
  constructor(
    @inject('GovernmentConstructionsLocalizationsRepository')
    private governmentConstructionsLocalizationsRepository: IGovernmentConstructionsLocalizationsRepository,

    @inject('GovernmentConstructionsLocalizationsProvider')
    private governmentConstructionsLocalizationProvicer: IGovernmentConstructionLocalizationProvider,
  ) {}

  public async execute(): Promise<void> {
    const result = await this.governmentConstructionsLocalizationProvicer.getAll();

    if (result !== undefined) {
      result.map((construction: ICreateConstructionsLocalizationDTO) =>
        this.governmentConstructionsLocalizationsRepository.create(
          construction,
        ),
      );
    }
  }
}

export default GetGovernnmentConstructionsService;
