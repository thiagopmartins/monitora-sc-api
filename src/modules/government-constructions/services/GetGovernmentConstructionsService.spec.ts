import FakeGovarnamentConstructionsLocalization from '../providers/fakes/FakeGovarnamentConstructionsLocalization';
import FakeConstructionsLocalizationRepository from '../repositories/fakes/FakeConstructionsLocalizationRepository';
import GetGovernnmentConstructionsService from './GetGovernmentConstructionsService';

describe('GetGovernnmentConstructionsService', () => {
  it('should be able to create a new construction localization', async () => {
    const fakeConstructionsLocalizationRepository = new FakeConstructionsLocalizationRepository();
    const fakeGovarnamentConstructionsLocalization = new FakeGovarnamentConstructionsLocalization();

    const getGovernmentConstructionsLocalization = new GetGovernnmentConstructionsService(
      fakeConstructionsLocalizationRepository,
      fakeGovarnamentConstructionsLocalization,
    );

    const create = jest.spyOn(
      fakeConstructionsLocalizationRepository,
      'create',
    );

    await getGovernmentConstructionsLocalization.execute();

    expect(create).toBeCalledWith(
      expect.objectContaining({
        construction_id: expect.any(String),
      }),
    );

    expect(
      fakeConstructionsLocalizationRepository.constructions,
    ).toMatchSnapshot([
      {
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        construction_id: expect.any(String),
        id: expect.any(String),
      },
    ]);
  });

  it('should be able to update if already exists construcion', async () => {
    const fakeConstructionsLocalizationRepository = new FakeConstructionsLocalizationRepository();
    const fakeGovarnamentConstructionsLocalization = new FakeGovarnamentConstructionsLocalization();

    const getGovernmentConstructionsLocalization = new GetGovernnmentConstructionsService(
      fakeConstructionsLocalizationRepository,
      fakeGovarnamentConstructionsLocalization,
    );

    const update = jest.spyOn(
      fakeConstructionsLocalizationRepository,
      'update',
    );

    await getGovernmentConstructionsLocalization.execute();
    await getGovernmentConstructionsLocalization.execute();

    expect(update).toBeCalledWith(
      expect.objectContaining({
        construction_id: expect.any(String),
      }),
    );

    expect(
      fakeConstructionsLocalizationRepository.constructions,
    ).toMatchSnapshot([
      {
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        construction_id: expect.any(String),
        id: expect.any(String),
      },
    ]);
  });

  it('should be not call repository functions if governmentConstructionsLocalizationProvider is undefined', async () => {
    const fakeConstructionsLocalizationRepository = new FakeConstructionsLocalizationRepository();
    const fakeGovarnamentConstructionsLocalization = new FakeGovarnamentConstructionsLocalization();

    const getGovernmentConstructionsLocalization = new GetGovernnmentConstructionsService(
      fakeConstructionsLocalizationRepository,
      fakeGovarnamentConstructionsLocalization,
    );

    jest
      .spyOn(fakeGovarnamentConstructionsLocalization, 'getAll')
      .mockImplementation(() => Promise.resolve(undefined));

    const find = jest.spyOn(
      fakeConstructionsLocalizationRepository,
      'findByConstructionLocalizarionId',
    );

    await getGovernmentConstructionsLocalization.execute();

    expect(find).not.toBeCalled();
  });
});
