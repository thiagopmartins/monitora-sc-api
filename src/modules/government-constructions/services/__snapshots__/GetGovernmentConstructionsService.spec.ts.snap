// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetGovernnmentConstructionsService should be able to create a new construction localization 1`] = `
Array [
  Object {
    "construction_id": Any<String>,
    "contracts": Array [
      Object {
        "constructionSitesId": Array [
          1,
        ],
        "id": 1026464,
      },
      Object {
        "constructionSitesId": Array [
          1,
        ],
        "id": 1027889,
      },
      Object {
        "constructionSitesId": Array [
          1,
        ],
        "id": 1029711,
      },
    ],
    "coordenates": Array [],
    "created_at": Any<Date>,
    "description": "PETROLÂNDIA (ENTR. SC-284) - ENTR. BR-282 (P/ BOCAINA DO SUL)",
    "heatMapQuantity": 2.12,
    "heatMapValue": 254912.79,
    "id": Any<String>,
    "nuLatitude": -27.535722175999,
    "nuLongitude": -49.698887383295,
    "projectsId": Array [
      2239,
      2259,
      3992,
    ],
    "updated_at": Any<Date>,
    "urlPin": "http://www.sicop.sc.gov.br:80/mapa-smo-backend/imagens/areas/345pin.png",
  },
]
`;

exports[`GetGovernnmentConstructionsService should be able to update if already exists construcion 1`] = `
Array [
  Object {
    "construction_id": Any<String>,
    "contracts": Array [
      Object {
        "constructionSitesId": Array [
          1,
        ],
        "id": 1026464,
      },
      Object {
        "constructionSitesId": Array [
          1,
        ],
        "id": 1027889,
      },
      Object {
        "constructionSitesId": Array [
          1,
        ],
        "id": 1029711,
      },
    ],
    "coordenates": Array [],
    "created_at": Any<Date>,
    "description": "PETROLÂNDIA (ENTR. SC-284) - ENTR. BR-282 (P/ BOCAINA DO SUL)",
    "heatMapQuantity": 2.12,
    "heatMapValue": 254912.79,
    "id": Any<String>,
    "nuLatitude": -27.535722175999,
    "nuLongitude": -49.698887383295,
    "projectsId": Array [
      2239,
      2259,
      3992,
    ],
    "updated_at": Any<Date>,
    "urlPin": "http://www.sicop.sc.gov.br:80/mapa-smo-backend/imagens/areas/345pin.png",
  },
]
`;
